src/librustc_typeck/lib.rs

```diff
@@ -188,10 +188,18 @@ fn check_main_fn_ty<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
                         hir::ItemFn(.., ref generics, _) => {
                             let mut error = false;
                             if !generics.params.is_empty() {
-                                struct_span_err!(tcx.sess, generics.span, E0131,
-                                    "`main` function is not allowed to have type parameters")
-                                    .span_label(generics.span,
-                                                "`main` cannot have type parameters")
+                                let param_type = if generics.is_lt_parameterized() {
+                                    "lifetime"
+                                } else {
+                                    "type"
+                                };
```

Comment:
Intrigued as to what the output is when you have `fn main<'a, T>() {}`.

Reply:
It would say lifetime, because one of the parameters is a lifetime. Would you go "lifetime nor type" parameters in such a case?